### ISSUE TITLE
[Rule Request] Weak Computed Property

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#2596]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

`weak` on a computed property doesn't do anything.

There [are](https://forums.swift.org/t/can-we-forbid-weak-keyword-for-computed-properties/19811) [discussions](https://github.com/apple/swift-evolution/pull/528) about adding a warning on the compiler directly.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
class A {
    var b1: B { return B() } // ok
    weak var b2: B? // ok

    weak var b3: B? { return nil } // warning
    weak var b4: B? {  // warning
        get { return b3 }
        set { b3 = newValue }
    }
}
```
3. Should the rule be configurable, if so what parameters should be configurable?

Just severity

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

I think it should be enabled by default. The only reasonable case for allowing this is to document the behavior, but since it's really not enforced by the compiler I'm not sure if it's worth the possible confusion.

**Comment by robertmryan** [comment]
I would suggest reopening this issue. 

Bottom line, there absolutely are situation where you want to use `weak` with computed properties, namely where the stored property backing this computed property is, itself, `weak`.

For example, consider:

```swift
class SomeClass {
    private weak var _delegate: SomeClassDelegate?
    var delegate: SomeClassDelegate? {             // whoops; this should be `weak`
        get { return _delegate }
        set { _delegate = newValue }
    }
}
```

And

```swift
class CustomDelegate: SomeClassDelegate { ... }
```

Then

```
let object = SomeClass()
object.delegate = CustomDelegate()
```

In the absence of the the `weak` qualifier on the computed property and without diving into the implementation details, the programmer might incorrectly conclude that the above is fine. But it’s not. Because the underlying `_delegate` is `weak`, this `CustomDelegate` instance will be deallocated immediately, and the object will end up with no delegate object. And there’s no compiler warning about this behavior.

If, however, we fix `SomeClass` like so:

```swift
class SomeClass {
    private weak var _delegate: SomeClassDelegate?
    weak var delegate: SomeClassDelegate? {        // great; matches underlying semantics
        get { return _delegate }
        set { _delegate = newValue }
    }
}
```

Then the programmer will receive a very helpful warning:

```swift
let object = SomeClass()
object.delegate = CustomDelegate()  // results in "Instance will be immediately deallocated because property 'delegate' is 'weak’"
```

They’ll then correctly deduce that they should keep their own strong reference to this CustomDelegate for the code to work properly.

So, bottom line, you don’t technically need the weak qualifier on the computed property that is backed by a private weak stored property, but it’s prudent to do so to avoid mysterious bugs and/or misunderstandings about the underlying semantics. 

It strikes me that the warning should be (a) if the getter is retrieving values that are not `weak`, then there should be an error if the computed property is marked `weak`; but (b) if the getter is retrieving values that _are_ `weak`, the the converse is true, that it should be an error if the computed property _is not_ marked `weak`. And if this level of complexity is not easily captured in SwiftLint, I might suggest disabling this rule altogether.

(https://stackoverflow.com/a/55348456/1271826)
